sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ConditionEvaluatesToConstant.cs

```diff
@@ -151,29 +150,36 @@ void explorationEnded(object sender, EventArgs args)
                         .Except(isUnknown)
                         .Except(isNotNull)
                         .Where(c => !IsInsideCatchOrFinallyBlock(c))
-                        .Select(node => Diagnostic.Create(s2589, node.GetLocation(), messageArgs: S2589MessageNull)))
+                        .Select(node => Diagnostic.Create(s2589, node.GetLocation(), S2589MessageNull)))
                     .Union(isNotNull
                         .Except(isUnknown)
                         .Except(isNull)
                         .Where(c => !IsInsideCatchOrFinallyBlock(c))
-                        .Select(node => Diagnostic.Create(s2583, node.GetLocation(), messageArgs: S2583MessageNotNull)))
-                    .ToList()
-                    .ForEach(d => context.ReportDiagnosticWhenActive(d));
+                        .Select(node => Diagnostic.Create(s2583, node.GetLocation(), S2583MessageNotNull)));
             }
 
-            explodedGraph.InstructionProcessed += instructionProcessed;
-            explodedGraph.ExplorationEnded += explorationEnded;
-            explodedGraph.ConditionEvaluated += collectConditions;
+            public void Dispose()
+            {
+                explodedGraph.InstructionProcessed -= InstructionProcessed;
+                explodedGraph.ConditionEvaluated -= ConditionEvaluatedHandler;
+            }
 
-            try
+            private void InstructionProcessed(object sender, InstructionProcessedEventArgs args)
             {
-                explodedGraph.Walk();
+                hasYieldStatement = hasYieldStatement || IsYieldNode(args.ProgramPoint.Block);
+                CollectCoalesce(args, isNull, isNotNull, isUnknown, context.SemanticModel);
             }
-            finally
+
+            private void ConditionEvaluatedHandler(object sender, ConditionEvaluatedEventArgs args) =>
+                CollectConditions(args, conditionTrue, conditionFalse, context.SemanticModel);
+
+            private static Diagnostic GetDiagnostics(SyntaxNode constantNode, bool constantValue)
```

Comment:
nitpick: you've moved the `GetDiagnostics` method here but I don't see a big need...

it has the same modifiers as `GetUnreachableLocations` which has been left where it was

just from a file history perspective. Also: I'm fine with leaving as it is if you think it improves the readability in any way.

Reply:
The problem was that the compiler will resolve `GetDiagnostics()` (the one without parameters) instead and I wanted to avoid a rename or a bigger refactoring by moving all the methods in the context. 

I've moved the method back and renamed it. I think it makes sense to stay near the other ones.